src/material/radio/radio.html

```diff
@@ -2,13 +2,17 @@
   <div class="mdc-radio" [class.mdc-radio--disabled]="disabled">
     <!-- Render this element first so the input is on top. -->
     <div class="mat-mdc-radio-touch-target" (click)="_onTouchTargetClick($event)"></div>
+    <!-- We don't want to add required attribute when button is used with MatRadioGroup as
+    we want to set directly it on MatRadioGroup as setting it on all MatRadioButton would be 
+    confusing for assistive technology users unless MatRadioButton is being used 
+    standalone or MatRadioButton is explicitly marked as required -->
     <input #input class="mdc-radio__native-control" type="radio"
            [id]="inputId"
            [checked]="checked"
            [disabled]="disabled && !disabledInteractive"
            [attr.name]="name"
            [attr.value]="value"
-           [required]="required"
+           [required]="radioGroup && required ? required : radioGroup ? null : required" 
```

Comment:
The nested ternary conditional statement is not easy to understand, especially without any parenthesis. This would be more readable if it just references a protected function like `getInputRequiredAttribute()`, and then you can also add that above comment as part of the function docs to explain the need for the nontrivial logic